include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,499 @@
+        {
+            emulated_float64_t lhs = createPreserveBitPattern(data);
+            emulated_float64_t rhsFlipped = rhs.flipSign();
+            
+            return lhs + rhsFlipped;
+        }
+
+        emulated_float64_t operator*(emulated_float64_t rhs)
+        {
+            emulated_float64_t retval = emulated_float64_t::createPreserveBitPattern(0u);
+
+            uint64_t lhsSign = data & ieee754::traits<float64_t>::signMask;
+            uint64_t rhsSign = rhs.data & ieee754::traits<float64_t>::signMask;
+            uint64_t lhsMantissa = ieee754::extractMantissa(data);
+            uint64_t rhsMantissa = ieee754::extractMantissa(rhs.data);
+            int lhsBiasedExp = ieee754::extractBiasedExponent(data);
+            int rhsBiasedExp = ieee754::extractBiasedExponent(rhs.data);
+
+            int exp = int(lhsBiasedExp + rhsBiasedExp) - ieee754::traits<float64_t>::exponentBias;
+            uint64_t sign = (data ^ rhs.data) & ieee754::traits<float64_t>::signMask;
+            if (!FastMath)
+            {
+                if (lhsBiasedExp == ieee754::traits<float64_t>::specialValueExp)
+                {
+                    if ((lhsMantissa != 0u) || ((rhsBiasedExp == ieee754::traits<float64_t>::specialValueExp) && (rhsMantissa != 0u)))
+                        return createPreserveBitPattern(impl::propagateFloat64NaN(data, rhs.data));
+                    if ((uint64_t(rhsBiasedExp) | rhsMantissa) == 0u)
+                        return createPreserveBitPattern(ieee754::traits<float64_t>::quietNaN);
+
+                    return createPreserveBitPattern(impl::assembleFloat64(sign, ieee754::traits<float64_t>::exponentMask, 0ull));
+                }
+                if (rhsBiasedExp == ieee754::traits<float64_t>::specialValueExp)
+                {
+                    /* a cannot be NaN, but is b NaN? */
+                    if (rhsMantissa != 0u)
+#ifdef RELAXED_NAN_PROPAGATION
+                        return rhs.data;
+#else
+                        return createPreserveBitPattern(impl::propagateFloat64NaN(data, rhs.data));
+#endif
+                    if ((uint64_t(lhsBiasedExp) | lhsMantissa) == 0u)
+                        return createPreserveBitPattern(ieee754::traits<float64_t>::quietNaN);
+
+                    return createPreserveBitPattern(sign | ieee754::traits<float64_t>::exponentMask);
+                }
+                if (lhsBiasedExp == 0)
+                {
+                    if (lhsMantissa == 0u)
+                        return createPreserveBitPattern(sign);
+                    impl::normalizeFloat64Subnormal(lhsMantissa, lhsBiasedExp, lhsMantissa);
+                }
+                if (rhsBiasedExp == 0)
+                {
+                    if (rhsMantissa == 0u)
+                        return createPreserveBitPattern(sign);
+                    impl::normalizeFloat64Subnormal(rhsMantissa, rhsBiasedExp, rhsMantissa);
+                }

```

Comment:
write some comments about what all this is doing

Reply:
it was copied from mesa, most of it has been removed